cloud-template/src/main/java/com/sequenceiq/cloudbreak/cloud/template/context/VolumeMatcher.java

```diff
@@ -0,0 +1,65 @@
+package com.sequenceiq.cloudbreak.cloud.template.context;
+
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.sequenceiq.cloudbreak.cloud.model.CloudInstance;
+import com.sequenceiq.cloudbreak.cloud.model.CloudResource;
+import com.sequenceiq.cloudbreak.cloud.model.VolumeSetAttributes;
+
+public class VolumeMatcher {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(VolumeMatcher.class);
+
+    private VolumeMatcher() {
+    }
+
+    public static void addVolumeResourcesToContext(List<CloudInstance> instances, List<CloudResource> groupInstances, List<CloudResource> groupVolumeSets,
+            ResourceBuilderContext context) {
+        List<CloudResource> groupVolumeSetsWithFQDN = groupVolumeSets.stream()
+                .filter(groupVolumeSet -> groupVolumeSet.getParameter("attributes", VolumeSetAttributes.class).getDiscoveryFQDN() != null)
+                .collect(Collectors.toList());
+        LOGGER.info("Volume sets with FQDN: " + groupVolumeSetsWithFQDN);
+        List<CloudResource> groupVolumeSetsWithoutFQDN = groupVolumeSets.stream()
+                .filter(groupVolumeSet -> groupVolumeSet.getParameter("attributes", VolumeSetAttributes.class).getDiscoveryFQDN() == null)
+                .collect(Collectors.toList());
```

Comment:
You could do the initial splitting of VolumeSets with/withOut FQDN with java streams partition. In high node count cases it might even perform better.

Reply:
Yes, it can be faster but it will cause a higher load on a single pod and in a multithreaded environment I am not a huge fan of stream partition because it will consume performance from other parallel tasks/ queries.

Reply:
I asked CP guys and it is okay to call CP readConfig on every client creation.